core/Testo.ml

```diff
@@ -191,79 +192,78 @@ let mask_line ?(mask = "<MASKED>") ?(after = "") ?(before = "") () =
-          assert (group_start >= match_start);
-          assert (group_stop <= match_stop);
-          let frag1 =
-            String.sub subj match_start (group_start - match_start) in
-          let to_be_replaced =
-            String.sub subj group_start (group_stop - group_start) in
-          let frag2 =
-            String.sub subj group_stop (match_stop - group_stop) in
-          frag1 ^ replace to_be_replaced ^ frag2
-    )
+         | `Text (* unmatched input *) str -> str
+         | `Delim (* match *) groups ->
+             let match_start, match_stop =
+               try Re.Group.offset groups 0 with Not_found -> assert false
+             in
+             let group_start, group_stop =
+               try Re.Group.offset groups 1
+               with Not_found -> (match_start, match_stop)
+             in
+             assert (group_start >= match_start);
+             assert (group_stop <= match_stop);
+             let frag1 =
+               String.sub subj match_start (group_start - match_start)
+             in
+             let to_be_replaced =
+               String.sub subj group_start (group_stop - group_start)
+             in
+             let frag2 = String.sub subj group_stop (match_stop - group_stop) in
+             frag1 ^ replace to_be_replaced ^ frag2)
     |> String.concat ""
 
 let path_segment_re = Re.Pcre.regexp {|[^\\/]+|}
 
 (* Internal function.
    Replace "/tmp/a/b" with "<TMP>/<MASKED>/<MASKED>" *)
-let default_replace_path ~tmpdir str =
-  let prefix_len = String.length tmpdir in
-  let len = String.length str in
-  assert (prefix_len <= len);
-  let suffix = String.sub str prefix_len (len - prefix_len) in
+(* Note: tmpdir_pat here since tmpdirs can be different paths on the same system *)
+let default_replace_path ~tmpdir_pat str =
+  let rex = Re.Pcre.regexp tmpdir_pat in
+  let suffix = Re.Pcre.substitute ~rex ~subst:(fun _ -> "") str in
   let new_suffix =
-    Re.Pcre.substitute
-      ~rex:path_segment_re
+    Re.Pcre.substitute ~rex:path_segment_re
       ~subst:(fun _seg -> "<MASKED>")
       suffix
   in
   "<TMP>" ^ new_suffix
 
-let mask_temp_paths
-    ?(depth = Some 1)
-    ?replace
-    ?(tmpdir = Filename.get_temp_dir_name ())
-    () =
+let mask_temp_paths ?(depth = Some 1) ?replace
+    ?(tmpdir = Filename.get_temp_dir_name ()) () =
+  (* Some OS's *cough cough MacOS* symlink their $TMPDIR to a different path *)
+  (* This means that if an ocmal program ever uses realpath, they may print a valid *)
+  (* But different temp dir that's not equal to $TMPDIR *)
+  (* So let's include that in the pattern *)
+  let real_tmpdir = Unix.realpath tmpdir in
+  let tmpdir =
+    if String.ends_with ~suffix:Filename.dir_sep tmpdir then
+      String.sub tmpdir 0 (String.length tmpdir - 1)
+    else tmpdir
+  in
   let tmpdir_pat = Re.Pcre.quote tmpdir in
+  let real_tmpdir_pat = Re.Pcre.quote real_tmpdir in
+  let tmpdir_pat = Printf.sprintf "(?:%s|%s)" tmpdir_pat real_tmpdir_pat in
```

Comment:
Looks good to me. One issue is that `Unix.realpath` requires OCaml >= 4.13 which is now 3 years old. It's fine if we really have to but not ideal. `String.ends_with` also requires OCaml >= 4.13 but is easier to reimplement ourselves.

If I understand correctly, there are two issues:
1. The trailing slash
2. Multiple temp folders

For (1), we need tests. In particular, we need to check that `"/tmp//"` works.

For (2), we also need tests. If depending on OCaml 4.13 is an issue, we could leave this responsibility to the user by letting them specify multiple temp folders. They can actually already do this today by applying two filters: one with the original path obtained with `Filename.get_temp_dir_name ()` and another that is `Filename.get_temp_dir_name () |> Unix.realpath`.

Reply:
Another issue: `String.sub tmpdir 0 (String.length tmpdir - 1)` assumes the length of `Filename.dir_sep` is 1 which is not guaranteed. Better use `String.length tmpdir - String.length Filename.dir_sep`.